opentelemetry/proto/collector/dynamicconfig/v1/dynamicconfig_service.proto

```diff
@@ -0,0 +1,159 @@
+// Copyright 2019, OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+syntax = "proto3";
+
+package opentelemetry.proto.collector.dynamicconfig.v1;
+
+import "opentelemetry/proto/resource/v1/resource.proto";
+
+option java_multiple_files = true;
+option java_package = "io.opentelemetry.proto.collector.dynamicconfig.v1";
+option java_outer_classname = "DynamicConfigServiceProto";
+option go_package = "github.com/open-telemetry/opentelemetry-proto/gen/go/collector/dynamicconfig/v1";
+
+// DynamicConfig is a service on the collector that enables updating metric
+// schedules, trace parameters, and other configurations on an agent/SDK without
+// having to restart the instrumented application. The collector may also act as
+// a bridge between third-party configuration services and the agent/sdk, piping
+// updated configs from a third-party source to an instrumented application.
+service DynamicConfig {
+  rpc GetConfig (ConfigRequest) returns (ConfigResponse);
+}
+
+message ConfigRequest{
+  // Describes the version of the config service protocol to use. Required.
+  int32 proto_version = 1;
+
+  // The target resource. This resource pings the config server requesting
+  // updated configs, and the server responds with the config information.
+  opentelemetry.proto.resource.v1.Resource resource = 2;
+
+  // Optional. Identifies the last configuration fingerprint / version number.
+  // The fingerprint serves as a quick way for the config server to determine
+  // whether an update to the configs had occurred -- if the client request's
+  // fingerprint differs from the server's, then the server knows that a change
+  // had occurred and an updated configuration should be sent.
+  string last_fingerprint = 3;
+}
+
+message ConfigResponse {
+  // Describes the version of the config service protocol to use. Should match
+  // the request’s version.
+  int32 proto_version = 1;
+
+  // Optional. The fingerprint associated with the current config settings. Each
+  // change in configs yields a different fingerprint. The client may cache a
+  // fingerprint, then return it on the next request to enable the server to
+  // quickly check whether a change in the configs occurred. If no changes
+  // occurred, then all other fields in the response are optional.
```

Comment:
Not sure if I follow, should nothing change, then the response will have only `fingerprint`, which equals the client's `last_known_fingerprint`, by which the client knows that no change happened? I'll also update the comments to make the process more clear

Reply:
Since a fingerprint is like a hash I don't think comparison with a previous value can guarantee nothing has changed due to hash collisions - it's more a hint which is more useful in the server which actually can compute the fingerprints.

So I think here for the client to know no change happened, it seems it needs to check the actual content. This would probably be many `hasX` checks using the generated code for proto and might work but may be indirect.

Another thought is perhaps this fingerprint is overoptimization, at least for now so maybe ok to remove for now. The configs don't seem so large and comparing fingerprints (etag + 204 responses in the http world) may not be worth it in practice.

Reply:
Great points, thanks for the quick response! I'll bring it up with the rest of the group, and we'll update our proposal (which will hopefully be made available soon)